File: python/prism_sdk/computational_execution_assurance.py
"""Python parity adapter for ``AFA-cli-P12-F28`` execution assurance."""
from __future__ import annotations

from dataclasses import dataclass, asdict
import hashlib
import json
import re
from typing import Any, Mapping

from .research_contracts import PRECLINICAL_BOUNDARY, RESEARCH_CONTRACT_SCHEMA_VERSION, ResearchContractError

FEATURE_ID = "AFA-cli-P12-F28"
CONTRACT_VERSION = "cli-federated-continual-computational-execution-assurance/1.0"
INPUT_SCHEMA = "ResearchWorkflowSpec4@1"
OUTPUT_SCHEMA = "ExecutionRun7@1"


def _hash(value: Any) -> str:
    return hashlib.sha256(json.dumps(value, sort_keys=True, separators=(",", ":"), ensure_ascii=False).encode()).hexdigest()


def _digest(value: Any) -> bool:
    return isinstance(value, str) and re.fullmatch(r"[0-9a-f]{64}", value) is not None


def _canonical(values: list[str] | tuple[str, ...]) -> bool:
    return list(values) == sorted(set(values))


@dataclass(frozen=True)
class ExecutionAssuranceReceipt:
    schema_version: str; contract_version: str; feature_id: str; request_id: str; run_id: str; workflow_id: str; scope: str; disposition: str
    plan_order: tuple[str, ...]; topological_order: tuple[str, ...]; completed_order: tuple[str, ...]; unresolved_order: tuple[str, ...]; blocked_order: tuple[str, ...]; cycle_order: tuple[str, ...]; missing_dependency_order: tuple[str, ...]; compensation_order: tuple[str, ...]; decisions: tuple[Mapping[str, Any], ...]; checkpoint_digest: str; replay_identity: str; run_digest: str; semantic_loss: tuple[Mapping[str, Any], ...]; omissions: tuple[str, ...]; uncertainty: tuple[str, ...]; negative_evidence: tuple[str, ...]; artifact: Mapping[str, Any]; effect_receipts: tuple[str, ...]; raw_data_local: bool; federation_export: str; boundary: str
    def validate(self) -> None:
        if (self.schema_version, self.contract_version, self.feature_id) != (RESEARCH_CONTRACT_SCHEMA_VERSION, CONTRACT_VERSION, FEATURE_ID) or self.boundary != PRECLINICAL_BOUNDARY or self.federation_export != "aggregate-digest-only" or self.raw_data_local is not True or not all(isinstance(value, str) and value.strip() for value in (self.request_id, self.run_id, self.workflow_id, self.scope)) or not self.plan_order or len(self.topological_order) > len(self.plan_order) or len(self.decisions) != len(self.plan_order) or not self.effect_receipts: raise ResearchContractError("execution identity, locality, plan, decisions, or effects are incomplete")
        for values in (self.plan_order, self.completed_order, self.unresolved_order, self.blocked_order, self.cycle_order, self.missing_dependency_order, self.compensation_order, self.omissions, self.uncertainty, self.negative_evidence, self.effect_receipts):
            if not _canonical(values): raise ResearchContractError("execution ordering is not canonical")
        if tuple(str(item.get("node_id", "")) for item in self.decisions) != self.plan_order: raise ResearchContractError("execution decisions do not match plan order")
        partition = list(self.completed_order) + list(self.unresolved_order) + list(self.blocked_order)
        if set(partition) != set(self.plan_order) or len(partition) != len(set(partition)): raise ResearchContractError("execution dispositions do not partition plan")
        if any(not effect.startswith("verify:execution-plan:") and effect != "block:unsafe-release" for effect in self.effect_receipts): raise ResearchContractError("execution effect is outside verification gate")
        if any(not _digest(value) for value in (self.checkpoint_digest, self.replay_identity, self.run_digest, self.artifact.get("content_hash"))): raise ResearchContractError("execution digest is invalid")
        if self.artifact.get("content_type") != "application/vnd.aurora.execution-run+json": raise ResearchContractError("execution artifact type is invalid")
    def to_dict(self) -> dict[str, Any]:
        value = asdict(self)
        for key, item in value.items():
            if isinstance(item, tuple): value[key] = list(item)
        return value


def assure_computational_execution(*, request: Mapping[str, Any]) -> ExecutionAssuranceReceipt:
    if any(not str(request.get(field, "")).strip() for field in ("request_id", "run_id", "workflow_id", "scope")) or request.get("plan_schema") != INPUT_SCHEMA or not request.get("nodes") or int(request.get("budget_units", 0)) <= 0 or int(request.get("budget_units", 0)) > int(request.get("max_budget_units", 0)) or request.get("raw_data_local") is not True or request.get("federated_summary_only") is not True or request.get("boundary") != PRECLINICAL_BOUNDARY or not _digest(request.get("replay_identity")): raise ResearchContractError("execution identity, bounds, locality, aggregate mode, replay, or boundary is invalid")
    nodes = sorted(request["nodes"], key=lambda item: str(item.get("node_id", ""))); ids = [str(node.get("node_id", "")) for node in nodes]
    if not all(ids) or len(set(ids)) != len(ids) or any(not _canonical([str(dep) for dep in node.get("dependency_order", [])]) for node in nodes): raise ResearchContractError("execution identifiers or dependency orders are invalid")
    plan_set = set(ids); indegree = {node_id: len(node.get("dependency_order", [])) for node_id, node in zip(ids, nodes)}; children: dict[str, list[str]] = {}
    for node in nodes:
        for dep in node.get("dependency_order", []): children.setdefault(str(dep), []).append(str(node["node_id"]))
    queue = sorted(node_id for node_id in ids if indegree[node_id] == 0); topological: list[str] = []
    while queue:
        node_id = queue.pop(0); topological.append(node_id)
        for child in sorted(children.get(node_id, [])):
            indegree[child] -= 1
            if indegree[child] == 0: queue.append(child); queue.sort()
    topological_set = set(topological); cycle = sorted(node_id for node in ids if node_id not in topological_set and all(str(dep) in plan_set for dep in next(item for item in nodes if str(item["node_id"]) == node_id).get("dependency_order", [])))
    global_failed = {gate for gate, failed in (("policy", not request.get("policy_allow", False)), ("protected-closure", not request.get("protected_closure", False)), ("raw-data-locality", request.get("raw_data_local") is not True), ("aggregate-only", request.get("federated_summary_only") is not True), ("adversarial-input", bool(request.get("adversarial_events")))) if failed}
    completed: list[str] = []; unresolved: list[str] = []; blocked: list[str] = []; missing: set[str] = set(); omissions: set[str] = set(); uncertainty: set[str] = set(); negative: set[str] = set(); decisions: list[dict[str, Any]] = []; semantic_loss: list[dict[str, Any]] = []; spent = 0; allowed = {"compute-local", "read-local", "write-artifact"}
    for node in nodes:
        node_id = str(node["node_id"]); failed = set(global_failed); pending: set[str] = set(); dependencies = [str(dep) for dep in node.get("dependency_order", [])]
        for dep in sorted(set(dependencies) - plan_set): pending.add("missing-dependency"); missing.add(f"{node_id}:{dep}"); omissions.add(f"{node_id}:missing-dependency:{dep}")
        if node_id in cycle: failed.add("dependency-cycle")
        if str(node.get("effect_kind", "")) not in allowed: failed.add("effect-not-local-allow-listed")
        if node.get("replay_identity") != request["replay_identity"]: failed.add("replay-identity")
        if not node.get("artifact_digest"): pending.add("artifact-digest-missing"); omissions.add(f"{node_id}:artifact-digest-missing")
        if not node.get("provenance_digest"): pending.add("provenance-missing"); omissions.add(f"{node_id}:provenance-missing")
        for item in node.get("omissions", []): pending.add("node-omissions"); omissions.add(f"{node_id}:{item}")
        for item in node.get("uncertainty", []): pending.add("node-uncertainty"); uncertainty.add(f"{node_id}:{item}")
        state = str(node.get("evidence_state", "unknown"));
        if state == "contradicted": failed.add("contradicted-evidence")
        elif state in {"unknown", "speculative"}: pending.add("evidence-state-not-qualified"); uncertainty.add(f"{node_id}:evidence-state")
        negative.add(f"{node_id}:execution-not-started"); cost = int(node.get("estimated_cost", 0))
        if cost > int(request["budget_units"]) - spent: pending.add("budget-ceiling"); omissions.add(f"{node_id}:budget-ceiling")
        else: spent += cost
        disposition = "blocked" if failed else "unresolved" if pending else "completed"; (blocked if disposition == "blocked" else unresolved if disposition == "unresolved" else completed).append(node_id); decisions.append({"node_id": node_id, "effect_kind": str(node.get("effect_kind", "")), "disposition": disposition, "failed_gates": sorted(failed), "conditional_gates": sorted(pending)})
        if failed: semantic_loss.append({"field": f"node:{node_id}", "reason": "execution node failed a release gate", "severity": "decision_relevant"})
    completed = sorted(set(completed)); unresolved = sorted(set(unresolved)); blocked = sorted(set(blocked)); compensation = tuple(sorted({f"retain:{node_id}:no-dispatch" for node_id in unresolved + blocked})); disposition = "blocked" if global_failed or blocked else "unresolved" if unresolved or len(completed) != len(ids) else "qualified"; checkpoint_digest = _hash({"run_id": request["run_id"], "workflow_id": request["workflow_id"], "plan_order": ids, "topological_order": topological, "replay_identity": request["replay_identity"]}); payload = {"schema_version": RESEARCH_CONTRACT_SCHEMA_VERSION, "contract_version": CONTRACT_VERSION, "feature_id": FEATURE_ID, "request_id": request["request_id"], "run_id": request["run_id"], "workflow_id": request["workflow_id"], "scope": request["scope"], "disposition": disposition, "plan_order": ids, "topological_order": topological, "completed_order": completed, "unresolved_order": unresolved, "blocked_order": blocked, "cycle_order": cycle, "missing_dependency_order": sorted(missing), "compensation_order": list(compensation), "decisions": decisions, "checkpoint_digest": checkpoint_digest, "replay_identity": request["replay_identity"], "boundary": PRECLINICAL_BOUNDARY}; run_digest = _hash(payload); artifact = {"schema_version": RESEARCH_CONTRACT_SCHEMA_VERSION, "artifact_id": f"execution-run:{request['run_id']}", "content_type": "application/vnd.aurora.execution-run+json", "content_hash": run_digest, "semantic_loss": semantic_loss, "provenance": [{"source_id": str(request["run_id"]), "relation": "computational-execution-assurance", "digest": run_digest}], "boundary": PRECLINICAL_BOUNDARY}; receipt = ExecutionAssuranceReceipt(RESEARCH_CONTRACT_SCHEMA_VERSION, CONTRACT_VERSION, FEATURE_ID, str(request["request_id"]), str(request["run_id"]), str(request["workflow_id"]), str(request["scope"]), disposition, tuple(ids), tuple(topological), tuple(completed), tuple(unresolved), tuple(blocked), tuple(cycle), tuple(sorted(missing)), compensation, tuple(decisions), checkpoint_digest, str(request["replay_identity"]), run_digest, tuple(semantic_loss), tuple(sorted(omissions)), tuple(sorted(uncertainty)), tuple(sorted(negative)), artifact, (f"verify:execution-plan:{request['run_id']}",) if disposition == "qualified" else ("block:unsafe-release",), True, "aggregate-digest-only", PRECLINICAL_BOUNDARY); receipt.validate(); return receipt


__all__ = ["FEATURE_ID", "CONTRACT_VERSION", "INPUT_SCHEMA", "OUTPUT_SCHEMA", "ExecutionAssuranceReceipt", "assure_computational_execution"]

File: crates/weavelang/src/computational_execution_assurance.rs
//! Local single-study computational-execution assurance for `AFA-weavelang-P12-F25`.
//!
//! This verifier consumes a caller-declared research workflow graph and emits a deterministic,
//! evidence-bearing execution-run envelope. It does not dispatch tools, contact networks, move
//! raw data, or claim that a scientific result is true. Unknown/unmeasured evidence, cycles,
//! missing dependencies, unauthorized effects, budget exhaustion, and adversarial events remain
//! explicit and fail closed.

use bioprism_foundation::{
    AutonomyTier, CapabilityManifest, Determinism, Effect, EvidenceReference, EvidenceState,
    ResearchSurface, TypedPort, TypedResearchArtifact, PRECLINICAL_BOUNDARY,
    RESEARCH_CONTRACT_SCHEMA_VERSION,
};
use bioprism_ids::ContentHash;
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::collections::{BTreeMap, BTreeSet};
use thiserror::Error;

pub const FEATURE_ID: &str = "AFA-weavelang-P12-F25";
pub const CONTRACT_VERSION: &str = "weavelang-local-computational-execution-assurance-harness/1.0";
pub const INPUT_SCHEMA: &str = "ResearchWorkflowSpec1@1";
pub const OUTPUT_SCHEMA: &str = "ExecutionRun7@1";
pub const TOOL_NAME: &str = "weavelang_computational_execution_assurance";
const CONTENT_TYPE: &str = "application/vnd.aurora.weavelang-execution-run-7+json";

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ExecutionEvidenceState {
    Proven,
    Supported,
    Unknown,
    Unmeasured,
    Contradicted,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ExecutionNode {
    pub node_id: String,
    pub capability_id: String,
    pub dependency_order: Vec<String>,
    pub effect: String,
    pub resource: String,
    pub cost: f64,
    pub evidence_state: ExecutionEvidenceState,
    pub input_digest: ContentHash,
    pub output_digest: ContentHash,
    pub permitted: bool,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ResearchWorkflowSpec {
    pub schema_version: String,
    pub request_id: String,
    pub workflow_id: String,
    pub actor_id: String,
    pub nodes: Vec<ExecutionNode>,
    pub budgets: BTreeMap<String, f64>,
    pub replay_identity: ContentHash,
    pub policy_allow: bool,
    pub protected_closure: bool,
    pub raw_data_local: bool,
    pub adversarial_events: Vec<String>,
    pub boundary: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ExecutionRunDisposition {
    Qualified,
    Unresolved,
    Blocked,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ExecutionRunReceipt {
    pub schema_version: String,
    pub contract_version: String,
    pub feature_id: String,
    pub request_id: String,
    pub workflow_id: String,
    pub actor_id: String,
    pub disposition: ExecutionRunDisposition,
    pub node_order: Vec<String>,
    pub selected_order: Vec<String>,
    pub unresolved_order: Vec<String>,
    pub blocked_order: Vec<String>,
    pub missing_dependency_order: Vec<String>,
    pub cycle_order: Vec<String>,
    pub unauthorized_effect_order: Vec<String>,
    pub budget_exhausted_order: Vec<String>,
    pub omission_order: Vec<String>,
    pub uncertainty_order: Vec<String>,
    pub negative_evidence_order: Vec<String>,
    pub replay_identity: ContentHash,
    pub run_digest: ContentHash,
    pub artifact: TypedResearchArtifact,
    pub effect_receipts: Vec<String>,
    pub raw_data_local: bool,
    pub boundary: String,
}

#[derive(Debug, Error, Clone, PartialEq, Eq)]
pub enum ComputationalExecutionError {
    #[error("invalid computational execution request: {0}")]
    Invalid(String),
    #[error("computational execution artifact failed: {0}")]
    Artifact(String),
    #[error("computational execution JSON failed: {0}")]
    Json(String),
}

fn invalid(message: impl Into<String>) -> ComputationalExecutionError {
    ComputationalExecutionError::Invalid(message.into())
}
fn digest_is_valid(value: &ContentHash) -> bool {
    value.as_str().len() == 64
}
fn canonical(values: &[String]) -> bool {
    values.windows(2).all(|pair| pair[0] < pair[1])
}

impl ExecutionRunReceipt {
    pub fn validate(&self) -> Result<(), ComputationalExecutionError> {
        if self.schema_version != RESEARCH_CONTRACT_SCHEMA_VERSION
            || self.contract_version != CONTRACT_VERSION
            || self.feature_id != FEATURE_ID
            || self.boundary != PRECLINICAL_BOUNDARY
            || !self.raw_data_local
            || self.request_id.trim().is_empty()
            || self.workflow_id.trim().is_empty()
            || self.actor_id.trim().is_empty()
            || self.node_order.is_empty()
            || self.effect_receipts.is_empty()
        {
            return Err(invalid(
                "execution identity, graph, locality, or effects are incomplete",
            ));
        }
        for values in [
            &self.node_order,
            &self.selected_order,
            &self.unresolved_order,
            &self.blocked_order,
            &self.missing_dependency_order,
            &self.cycle_order,
            &self.unauthorized_effect_order,
            &self.budget_exhausted_order,
            &self.omission_order,
            &self.uncertainty_order,
            &self.negative_evidence_order,
            &self.effect_receipts,
        ] {
            if !canonical(values) {
                return Err(invalid("execution ordering is not canonical"));
            }
        }
        let ids = self.node_order.iter().cloned().collect::<BTreeSet<_>>();
        let parts = self
            .selected_order
            .iter()
            .chain(self.unresolved_order.iter())
            .chain(self.blocked_order.iter())
            .cloned()
            .collect::<Vec<_>>();
        if parts.len() != ids.len() || parts.iter().cloned().collect::<BTreeSet<_>>() != ids {
            return Err(invalid("execution node states do not partition the graph"));
        }
        for value in [
            &self.replay_identity,
            &self.run_digest,
            &self.artifact.content_hash,
        ] {
            if !digest_is_valid(value) {
                return Err(invalid("execution digest is invalid"));
            }
        }
        self.artifact
            .validate_metadata()
            .map_err(|error| ComputationalExecutionError::Artifact(error.to_string()))?;
        if self.artifact.content_type != CONTENT_TYPE {
            return Err(invalid("execution artifact type is invalid"));
        }
        if self.disposition == ExecutionRunDisposition::Qualified
            && self.effect_receipts != [format!("verify:weavelang-execution:{}", self.workflow_id)]
        {
            return Err(invalid("qualified execution effect is invalid"));
        }
        if self.disposition != ExecutionRunDisposition::Qualified
            && self.effect_receipts != ["block:unsafe-release"]
        {
            return Err(invalid("non-qualified execution must block release"));
        }
        Ok(())
    }
}

pub fn computational_execution_assurance_manifest() -> CapabilityManifest {
    CapabilityManifest {
        schema_version: RESEARCH_CONTRACT_SCHEMA_VERSION.into(),
        capability_id: FEATURE_ID.into(),
        version: CONTRACT_VERSION.into(),
        owner_crate: "weavelang".into(),
        consumers: BTreeSet::from([
            String::from("platform reliability engineer"),
            String::from("research workflow operator"),
            String::from("downstream execution gateway"),
        ]),
        behavior: "verifies a bounded local research execution graph, evidence closure, effect allow-list, replay identity, and resource budget without dispatching any effect".into(),
        value: "turns a WeaveLang execution plan into a deterministic ExecutionRun7 assurance artifact while making missing evidence, cycles, and unsafe effects impossible to mistake for success".into(),
        inputs: vec![TypedPort { name: "research_workflow_spec".into(), schema: INPUT_SCHEMA.into(), required: true }],
        outputs: vec![TypedPort { name: "execution_run".into(), schema: OUTPUT_SCHEMA.into(), required: true }],
        effects: BTreeSet::from([Effect::ReadLocalData, Effect::ExecuteLocalComputation, Effect::WriteLocalArtifact]),
        permissions: BTreeSet::from([String::from("evaluate:capability-runs")]),
        determinism: Determinism::ByteStable,
        evidence: vec![
            EvidenceReference { source_id: "cwl".into(), state: EvidenceState::Supported, locator: Some("https://www.commonwl.org/specification/".into()) },
            EvidenceReference { source_id: "ga4gh-wes".into(), state: EvidenceState::Supported, locator: Some("https://ga4gh.github.io/workflow-execution-service-schemas/docs/".into()) },
            EvidenceReference { source_id: "slsa-provenance-1.2".into(), state: EvidenceState::Supported, locator: Some("https://slsa.dev/spec/v1.2/provenance".into()) },
        ],
        authority_requirements: Vec::new(),
        autonomy_tier: AutonomyTier::A0,
        surfaces: BTreeSet::from([ResearchSurface::Ui, ResearchSurface::Api, ResearchSurface::Sdk, ResearchSurface::McpTool, ResearchSurface::Protocol, ResearchSurface::Policy, ResearchSurface::Operator]),
        boundary: PRECLINICAL_BOUNDARY.into(),
    }
}

pub fn assure_computational_execution(
    request: &ResearchWorkflowSpec,
) -> Result<ExecutionRunReceipt, ComputationalExecutionError> {
    validate_request(request)?;
    let mut nodes = request.nodes.clone();
    nodes.sort_by(|a, b| a.node_id.cmp(&b.node_id));
    let node_order = nodes
        .iter()
        .map(|node| node.node_id.clone())
        .collect::<Vec<_>>();
    let ids = node_order.iter().cloned().collect::<BTreeSet<_>>();
    let mut missing = BTreeSet::new();
    let mut cycles = BTreeSet::new();
    let mut unauthorized = BTreeSet::new();
    let mut budget_exhausted = BTreeSet::new();
    let mut selected = BTreeSet::new();
    let mut unresolved = BTreeSet::new();
    let mut blocked = BTreeSet::new();
    let mut omissions = BTreeSet::new();
    let mut uncertainty = BTreeSet::new();
    let mut negative = BTreeSet::new();
    let allowed = [
        "read-local",
        "execute-local-computation",
        "write-local-artifact",
    ];
    for node in &nodes {
        for dependency in &node.dependency_order {
            if !ids.contains(dependency) {
                missing.insert(format!("{}->{dependency}", node.node_id));
            }
        }
        if !allowed.contains(&node.effect.as_str()) || !node.permitted {
            unauthorized.insert(node.node_id.clone());
        }
        if node.cost.is_sign_negative()
            || !node.cost.is_finite()
            || request.budgets.get(&node.resource).copied().unwrap_or(0.0) < node.cost
        {
            budget_exhausted.insert(node.node_id.clone());
        }
        match node.evidence_state {
            ExecutionEvidenceState::Contradicted => {
                blocked.insert(node.node_id.clone());
                negative.insert(format!("{}:contradicted", node.node_id));
            }
            ExecutionEvidenceState::Unknown => {
                unresolved.insert(node.node_id.clone());
                uncertainty.insert(format!("{}:unknown", node.node_id));
            }
            ExecutionEvidenceState::Unmeasured => {
                unresolved.insert(node.node_id.clone());
                omissions.insert(format!("{}:unmeasured", node.node_id));
            }
            ExecutionEvidenceState::Proven | ExecutionEvidenceState::Supported => {}
        }
    }
    let mut marks = BTreeMap::<String, u8>::new();
    fn visit(
        node: &str,
        by_id: &BTreeMap<String, ExecutionNode>,
        marks: &mut BTreeMap<String, u8>,
        cycles: &mut BTreeSet<String>,
    ) {
        if marks.get(node) == Some(&1) {
            cycles.insert(node.to_string());
            return;
        }
        if marks.get(node) == Some(&2) {
            return;
        }
        marks.insert(node.to_string(), 1);
        if let Some(value) = by_id.get(node) {
            for dependency in &value.dependency_order {
                if by_id.contains_key(dependency) {
                    visit(dependency, by_id, marks, cycles);
                }
            }
        }
        marks.insert(node.to_string(), 2);
    }
    let by_id = nodes
        .iter()
        .cloned()
        .map(|node| (node.node_id.clone(), node))
        .collect::<BTreeMap<_, _>>();
    for id in &node_order {
        visit(id, &by_id, &mut marks, &mut cycles);
    }
    let global_block = !request.policy_allow
        || !request.protected_closure
        || !request.raw_data_local
        || !request.adversarial_events.is_empty();
    if !request.policy_allow {
        negative.insert("request:policy-denied".into());
    }
    if !request.protected_closure {
        uncertainty.insert("request:protected-closure-incomplete".into());
    }
    if !request.raw_data_local {
        negative.insert("request:raw-data-locality-violation".into());
    }
    negative.extend(
        request
            .adversarial_events
            .iter()
            .map(|event| format!("adversarial:{event}")),
    );
    blocked.extend(unauthorized.iter().cloned());
    blocked.extend(budget_exhausted.iter().cloned());
    blocked.extend(cycles.iter().cloned());
    blocked.extend(
        missing
            .iter()
            .filter_map(|value| value.split("->").next().map(str::to_string)),
    );
    if !cycles.is_empty() {
        blocked.extend(node_order.iter().cloned());
    }
    if global_block {
        blocked.extend(node_order.iter().cloned());
        selected.clear();
        unresolved.clear();
        omissions.insert("request:weavelang-release-gate-blocked".into());
    }
    if !global_block {
        selected.extend(
            node_order
                .iter()
                .filter(|id| !blocked.contains(*id) && !unresolved.contains(*id))
                .cloned(),
        );
    }
    let disposition = if global_block || !blocked.is_empty() {
        ExecutionRunDisposition::Blocked
    } else if !unresolved.is_empty() || !missing.is_empty() {
        ExecutionRunDisposition::Unresolved
    } else {
        ExecutionRunDisposition::Qualified
    };
    let selected_order = selected.into_iter().collect::<Vec<_>>();
    let unresolved_order = unresolved.into_iter().collect::<Vec<_>>();
    let blocked_order = blocked.into_iter().collect::<Vec<_>>();
    let missing_dependency_order = missing.into_iter().collect::<Vec<_>>();
    let cycle_order = cycles.into_iter().collect::<Vec<_>>();
    let unauthorized_effect_order = unauthorized.into_iter().collect::<Vec<_>>();
    let budget_exhausted_order = budget_exhausted.into_iter().collect::<Vec<_>>();
    let omission_order = omissions.into_iter().collect::<Vec<_>>();
    let uncertainty_order = uncertainty.into_iter().collect::<Vec<_>>();
    let negative_evidence_order = negative.into_iter().collect::<Vec<_>>();
    let effect_receipts = if disposition == ExecutionRunDisposition::Qualified {
        vec![format!(
            "verify:weavelang-execution:{}",
            request.workflow_id
        )]
    } else {
        vec!["block:unsafe-release".into()]
    };
    let payload = json!({"schema_version": RESEARCH_CONTRACT_SCHEMA_VERSION, "contract_version": CONTRACT_VERSION, "feature_id": FEATURE_ID, "request_id": request.request_id, "workflow_id": request.workflow_id, "actor_id": request.actor_id, "disposition": disposition, "node_order": node_order, "selected_order": selected_order, "unresolved_order": unresolved_order, "blocked_order": blocked_order, "missing_dependency_order": missing_dependency_order, "cycle_order": cycle_order, "unauthorized_effect_order": unauthorized_effect_order, "budget_exhausted_order": budget_exhausted_order, "omission_order": omission_order, "uncertainty_order": uncertainty_order, "negative_evidence_order": negative_evidence_order, "replay_identity": request.replay_identity, "effect_receipts": effect_receipts, "raw_data_local": request.raw_data_local, "boundary": PRECLINICAL_BOUNDARY});
    let run_digest = ContentHash::of_value(&payload)
        .map_err(|error| ComputationalExecutionError::Artifact(error.to_string()))?;
    let artifact = TypedResearchArtifact::from_payload(
        format!("weavelang-execution-run:{}", request.workflow_id),
        CONTENT_TYPE,
        &payload,
        Vec::new(),
        Vec::new(),
    )
    .map_err(|error| ComputationalExecutionError::Artifact(error.to_string()))?;
    let receipt = ExecutionRunReceipt {
        schema_version: RESEARCH_CONTRACT_SCHEMA_VERSION.into(),
        contract_version: CONTRACT_VERSION.into(),
        feature_id: FEATURE_ID.into(),
        request_id: request.request_id.clone(),
        workflow_id: request.workflow_id.clone(),
        actor_id: request.actor_id.clone(),
        disposition,
        node_order: payload["node_order"]
            .as_array()
            .unwrap()
            .iter()
            .map(|v| v.as_str().unwrap().into())
            .collect(),
        selected_order: payload["selected_order"]
            .as_array()
            .unwrap()
            .iter()
            .map(|v| v.as_str().unwrap().into())
            .collect(),
        unresolved_order: payload["unresolved_order"]
            .as_array()
            .unwrap()
            .iter()
            .map(|v| v.as_str().unwrap().into())
            .collect(),
        blocked_order: payload["blocked_order"]
            .as_array()
            .unwrap()
            .iter()
            .map(|v| v.as_str().unwrap().into())
            .collect(),
        missing_dependency_order: payload["missing_dependency_order"]
            .as_array()
            .unwrap()
            .iter()
            .map(|v| v.as_str().unwrap().into())
            .collect(),
        cycle_order: payload["cycle_order"]
            .as_array()
            .unwrap()
            .iter()
            .map(|v| v.as_str().unwrap().into())
            .collect(),
        unauthorized_effect_order: payload["unauthorized_effect_order"]
            .as_array()
            .unwrap()
            .iter()
            .map(|v| v.as_str().unwrap().into())
            .collect(),
        budget_exhausted_order: payload["budget_exhausted_order"]
            .as_array()
            .unwrap()
            .iter()
            .map(|v| v.as_str().unwrap().into())
            .collect(),
        omission_order: payload["omission_order"]
            .as_array()
            .unwrap()
            .iter()
            .map(|v| v.as_str().unwrap().into())
            .collect(),
        uncertainty_order: payload["uncertainty_order"]
            .as_array()
            .unwrap()
            .iter()
            .map(|v| v.as_str().unwrap().into())
            .collect(),
        negative_evidence_order: payload["negative_evidence_order"]
            .as_array()
            .unwrap()
            .iter()
            .map(|v| v.as_str().unwrap().into())
            .collect(),
        replay_identity: request.replay_identity.clone(),
        run_digest,
        artifact,
        effect_receipts: payload["effect_receipts"]
            .as_array()
            .unwrap()
            .iter()
            .map(|v| v.as_str().unwrap().into())
            .collect(),
        raw_data_local: request.raw_data_local,
        boundary: PRECLINICAL_BOUNDARY.into(),
    };
    receipt.validate()?;
    Ok(receipt)
}

pub fn assure_computational_execution_json(value: &Value) -> Result<Value, String> {
    let request: ResearchWorkflowSpec = serde_json::from_value(value.clone())
        .map_err(|error| format!("invalid research workflow spec: {error}"))?;
    let receipt = assure_computational_execution(&request).map_err(|error| error.to_string())?;
    serde_json::to_value(receipt)
        .map_err(|error| format!("cannot serialize execution run: {error}"))
}

pub fn validate_computational_execution_json(value: &Value) -> Result<ExecutionRunReceipt, String> {
    let receipt: ExecutionRunReceipt = serde_json::from_value(value.clone())
        .map_err(|error| format!("invalid execution run receipt: {error}"))?;
    receipt.validate().map_err(|error| error.to_string())?;
    Ok(receipt)
}

fn validate_request(request: &ResearchWorkflowSpec) -> Result<(), ComputationalExecutionError> {
    if request.schema_version != RESEARCH_CONTRACT_SCHEMA_VERSION
        || request.request_id.trim().is_empty()
        || request.workflow_id.trim().is_empty()
        || request.actor_id.trim().is_empty()
        || request.nodes.is_empty()
        || !digest_is_valid(&request.replay_identity)
        || !canonical(&request.adversarial_events)
        || !request.raw_data_local
        || request.boundary != PRECLINICAL_BOUNDARY
    {
        return Err(invalid(
            "workflow identity, graph, replay, locality, or boundary is invalid",
        ));
    }
    if request
        .budgets
        .values()
        .any(|value| !value.is_finite() || value.is_sign_negative())
    {
        return Err(invalid("workflow budgets must be finite and non-negative"));
    }
    let mut ids = BTreeSet::new();
    for node in &request.nodes {
        if node.node_id.trim().is_empty()
            || !ids.insert(node.node_id.clone())
            || node.capability_id.trim().is_empty()
            || node.effect.trim().is_empty()
            || node.resource.trim().is_empty()
            || !canonical(&node.dependency_order)
            || !digest_is_valid(&node.input_digest)
            || !digest_is_valid(&node.output_digest)
        {
            return Err(invalid(format!(
                "workflow node {} is malformed or duplicated",
                node.node_id
            )));
        }
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    fn hash(value: &str) -> ContentHash {
        ContentHash::of_bytes(value.as_bytes())
    }
    fn request() -> ResearchWorkflowSpec {
        let h = hash("execution");
        ResearchWorkflowSpec {
            schema_version: RESEARCH_CONTRACT_SCHEMA_VERSION.into(),
            request_id: "exec:one".into(),
            workflow_id: "workflow:one".into(),
            actor_id: "operator:one".into(),
            nodes: vec![ExecutionNode {
                node_id: "node:a".into(),
                capability_id: "cap:a".into(),
                dependency_order: vec![],
                effect: "execute-local-computation".into(),
                resource: "cpu".into(),
                cost: 1.0,
                evidence_state: ExecutionEvidenceState::Supported,
                input_digest: h.clone(),
                output_digest: h.clone(),
                permitted: true,
            }],
            budgets: BTreeMap::from([(String::from("cpu"), 2.0)]),
            replay_identity: h,
            policy_allow: true,
            protected_closure: true,
            raw_data_local: true,
            adversarial_events: vec![],
            boundary: PRECLINICAL_BOUNDARY.into(),
        }
    }
    #[test]
    fn manifest_is_a0() {
        assert_eq!(
            computational_execution_assurance_manifest().autonomy_tier,
            AutonomyTier::A0
        );
    }
    #[test]
    fn qualified() {
        assert_eq!(
            assure_computational_execution(&request())
                .unwrap()
                .disposition,
            ExecutionRunDisposition::Qualified
        );
    }
    #[test]
    fn deterministic() {
        let a = assure_computational_execution(&request()).unwrap();
        let b = assure_computational_execution(&request()).unwrap();
        assert_eq!(a.run_digest, b.run_digest);
    }
    #[test]
    fn unknown_is_unresolved() {
        let mut value = request();
        value.nodes[0].evidence_state = ExecutionEvidenceState::Unknown;
        assert_eq!(
            assure_computational_execution(&value).unwrap().disposition,
            ExecutionRunDisposition::Unresolved
        );
    }
    #[test]
    fn unauthorized_effect_blocks() {
        let mut value = request();
        value.nodes[0].effect = "external-network".into();
        assert_eq!(
            assure_computational_execution(&value).unwrap().disposition,
            ExecutionRunDisposition::Blocked
        );
    }
    #[test]
    fn cycle_blocks() {
        let mut value = request();
        value.nodes.push(ExecutionNode {
            node_id: "node:b".into(),
            capability_id: "cap:b".into(),
            dependency_order: vec!["node:a".into()],
            effect: "execute-local-computation".into(),
            resource: "cpu".into(),
            cost: 1.0,
            evidence_state: ExecutionEvidenceState::Supported,
            input_digest: hash("b"),
            output_digest: hash("b"),
            permitted: true,
        });
        value.nodes[0].dependency_order = vec!["node:b".into()];
        assert_eq!(
            assure_computational_execution(&value).unwrap().disposition,
            ExecutionRunDisposition::Blocked
        );
    }
    #[test]
    fn adversarial_blocks() {
        let mut value = request();
        value.adversarial_events = vec!["poisoned-artifact".into()];
        assert_eq!(
            assure_computational_execution(&value).unwrap().disposition,
            ExecutionRunDisposition::Blocked
        );
    }
}
